topic_manager.go

```diff
@@ -105,79 +113,175 @@ func (m *topicManager) Close() error {
+		value := v
+		topicDetail.ConfigEntries[k] = &value
 	}
-	response, err := m.broker.CreateTopics(&request)
 
+	err := m.admin.CreateTopic(topic, topicDetail, false)
 	if err != nil {
-		var errs []string
-		for k, topicErr := range response.TopicErrors {
-			errs = append(errs, fmt.Sprintf("%s: %s (%v)", k, topicErr.Err.Error(), topicErr.ErrMsg))
-		}
-		return fmt.Errorf("error creating topic %s, npar=%d, rfactor=%d, config=%#v: %v\ntopic errors:\n%s",
-			topic, npar, rfactor, config, err, strings.Join(errs, "\n"))
+		return fmt.Errorf("error creating topic %s, npar=%d, rfactor=%d, config=%#v: %v",
+			topic, npar, rfactor, config, err)
 	}
 
 	return nil
 }
 
+func (m *topicManager) handleConfigMismatch(message string) error {
+	switch m.topicManagerConfig.MismatchBehavior {
+	case TMConfigMismatchBehaviorWarn:
+		m.topicManagerConfig.Logger.Printf("Warning: %s", message)
+		return nil
+	case TMConfigMismatchBehaviorFail:
+		return fmt.Errorf("%s", message)
+	}
+	// default is to ignore
+	return nil
+}
+
 func (m *topicManager) ensureExists(topic string, npar, rfactor int, config map[string]string) error {
-	exists, err := m.checkTopicExistsWithPartitions(topic, npar)
+
+	var hasTopic bool
+	partitions, err := m.Partitions(topic)
+
 	if err != nil {
-		return fmt.Errorf("error checking topic exists: %v", err)
+		if err != errTopicNotFound {
+			return fmt.Errorf("error checking topic: %v", err)
+		}
+	}
+
+	if len(partitions) > 0 {
+		hasTopic = true
+	}
+
+	// partitions do not match
+	if hasTopic && len(partitions) != npar {
+		return m.handleConfigMismatch(fmt.Sprintf("partition count mismatch for topic %s. Need %d, but existing topic has %d", topic, npar, len(partitions)))
+	}
+
+	// check additional config values via the cluster admin if our current version supports it
+	if hasTopic && m.adminSupported() {
+		cfgMap, err := m.getTopicConfigMap(topic)
+		if err != nil {
+			return err
+		}
+
+		// check for all user-passed config values whether they're as expected
+		for key, value := range config {
+			entry, ok := cfgMap[key]
+			if !ok {
+				return m.handleConfigMismatch(fmt.Sprintf("config for topic %s did not contain requested key %s", topic, key))
+			}
+			if entry.Value != value {
+				return m.handleConfigMismatch(fmt.Sprintf("unexpected config value for topic %s. Expected %s=%s. Got %s=%s", topic, key, value, key, entry.Value))
+			}
+		}
+
+		// check if the number of replicas match what we expect
+		topicMinReplicas, err := m.getTopicMinReplicas(topic)
+		if err != nil {
+			return err
+		}
+		if topicMinReplicas != rfactor {
+			return m.handleConfigMismatch(fmt.Sprintf("unexpected replication factor for topic %s. Expected %d, got %d",
+				topic,
+				rfactor,
+				topicMinReplicas))
+		}
 	}
-	if exists {
+
+	// already exists, no need to do create
+	if hasTopic {
 		return nil
 	}
+
 	return m.createTopic(topic,
 		npar,
 		rfactor,
 		config)
 }
```

Comment:
can we reduce the nested `if`s and return early, like this?:
```
if !hasTopic {
    return createTopic
}

if !m.adminSupported() {
    return nil
}

...
configMismatch code
...

```

Reply:
or do i miss something?

Reply:
nope, you're right, it was kind of messy. Tried to improve it